In [32]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [33]:
# Importing Data
path = r'C:\Users\arafo\CareerFoundry\Hotel Booking Analysis'
df_bookings = pd.read_csv(os.path.join(path, '02 - Data', 'Original Data', 'hotel_bookings.csv'), index_col = False)
df_countries = pd.read_csv(os.path.join(path, '02 - Data', 'Original Data', 'country_code.csv'), index_col = False)

In [34]:
pd.options.display.max_rows = None
pd.options.display.max_columns = None

In [35]:
df_bookings.head(15)

hotel  is_canceled  lead_time  arrival_date_year  \
0   Resort Hotel            0        342               2015   
1   Resort Hotel            0        737               2015   
2   Resort Hotel            0          7               2015   
3   Resort Hotel            0         13               2015   
4   Resort Hotel            0         14               2015   
5   Resort Hotel            0         14               2015   
6   Resort Hotel            0          0               2015   
7   Resort Hotel            0          9               2015   
8   Resort Hotel            1         85               2015   
9   Resort Hotel            1         75               2015   
10  Resort Hotel            1         23               2015   
11  Resort Hotel            0         35               2015   
12  Resort Hotel            0         68               2015   
13  Resort Hotel            0         18               2015   
14  Resort Hotel            0         37               2015   

   arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0                July                        27                          1   
1                July                        27                          1   
2                July                        27                          1   
3                July                        27                          1   
4                July                        27                          1   
5                July                        27                          1   
6                July                        27                          1   
7                July                        27                          1   
8                July                        27                          1   
9                July                        27                          1   
10               July                        27                          1   
11               July                        27                          1   
12               July                        27                          1   
13               July                        27                          1   
14               July                        27                          1   

    stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                         0                     0       2       0.0       0   
1                         0                     0       2       0.0       0   
2                         0                     1       1       0.0       0   
3                         0                     1       1       0.0       0   
4                         0                     2       2       0.0       0   
5                         0                     2       2       0.0       0   
6                         0                     2       2       0.0       0   
7                         0                     2       2       0.0       0   
8                         0                     3       2       0.0       0   
9                         0                     3       2       0.0       0   
10                        0                     4       2       0.0       0   
11                        0                     4       2       0.0       0   
12                        0                     4       2       0.0       0   
13                        0                     4       2       1.0       0   
14                        0                     4       2       0.0       0   

   meal country market_segment distribution_channel  is_repeated_guest  \
0    BB     PRT         Direct               Direct                  0   
1    BB     PRT         Direct               Direct                  0   
2    BB     GBR         Direct               Direct                  0   
3    BB     GBR      Corporate            Corporate                  0   
4    BB     GBR      Online TA                TA/TO                  0   
5    BB     GBR      Online TA              

In [36]:
df_bookings.tail(15)

hotel  is_canceled  lead_time  arrival_date_year  \
119375  City Hotel            0        200               2017   
119376  City Hotel            0        165               2017   
119377  City Hotel            0        185               2017   
119378  City Hotel            0        247               2017   
119379  City Hotel            0        109               2017   
119380  City Hotel            0         44               2017   
119381  City Hotel            0        188               2017   
119382  City Hotel            0        135               2017   
119383  City Hotel            0        164               2017   
119384  City Hotel            0         21               2017   
119385  City Hotel            0         23               2017   
119386  City Hotel            0        102               2017   
119387  City Hotel            0         34               2017   
119388  City Hotel            0        109               2017   
119389  City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
119375             August                        35   
119376             August                        35   
119377             August                        35   
119378             August                        35   
119379             August                        35   
119380             August                        35   
119381             August                        35   
119382             August                        35   
119383             August                        35   
119384             August                        35   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
119375                         30                        1   
119376                         29                        1   
119377                         30                        1   
119378                         31                        1   
119379                         31                        1   
119380                         31                        1   
119381                         31                        2   
119382                         30                        2   
119383                         31                        2   
119384                         30                        2   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  children  babies meal country  \
119375                     4       2       0.0       0   BB     AUT   
119376                     5       3       0.0       0   BB     AUT   
119377                     4       2       0.0       0   SC     CHE   
119378                     3       2       0.0       0   BB     GBR   
119379                     3       2       0.0       0   BB     GBR   
119380                     3       2       0.0       0   SC     DEU   
119381                     3       2       0.0       0   BB     DEU   
119382                     4       3       0.0       0   BB     JPN   
119383                     4       2       0.0       0   BB     DEU   
119384                     5       2       0.0       0   BB     BEL   
119385                     5       2       0.0       0   BB     BEL   
119386                     5       3       0.0       0   BB     FRA   
119387                     5       2       0.0       0   BB     DEU   
119388                     5       2       0.0       0   BB     GBR   
119389                     7     

In [37]:
# Checking countries of origins
df_bookings['country'].value_counts(dropna = False)

PRT    48590
GBR    12129
FRA    10415
ESP     8568
DEU     7287
ITA     3766
IRL     3375
BEL     2342
BRA     2224
NLD     2104
USA     2097
CHE     1730
CN      1279
AUT     1263
SWE     1024
CHN      999
POL      919
ISR      669
RUS      632
NOR      607
ROU      500
NaN      488
FIN      447
DNK      435
AUS      426
AGO      362
LUX      287
MAR      259
TUR      248
HUN      230
ARG      214
JPN      197
CZE      171
IND      152
KOR      133
GRC      128
DZA      103
SRB      101
HRV      100
MEX       85
EST       83
IRN       83
LTU       81
ZAF       80
BGR       75
NZL       74
COL       71
UKR       68
MOZ       67
CHL       65
SVK       65
THA       59
SVN       57
ISL       57
LVA       55
ARE       51
CYP       51
TWN       51
SAU       48
PHL       40
TUN       39
SGP       39
IDN       35
NGA       34
EGY       32
URY       32
LBN       31
PER       29
HKG       29
MYS       28
ECU       27
VEN       26
BLR       26
CPV       24
GEO       22
JOR       21
KAZ       19

In [38]:
df_countries

name alpha-2 alpha-3  \
0                                     Afghanistan      AF     AFG   
1                                   Åland Islands      AX     ALA   
2                                         Albania      AL     ALB   
3                                         Algeria      DZ     DZA   
4                                  American Samoa      AS     ASM   
5                                         Andorra      AD     AND   
6                                          Angola      AO     AGO   
7                                        Anguilla      AI     AIA   
8                                      Antarctica      AQ     ATA   
9                             Antigua and Barbuda      AG     ATG   
10                                      Argentina      AR     ARG   
11                                        Armenia      AM     ARM   
12                                          Aruba      AW     ABW   
13                                      Australia      AU     AUS   
14                                        Austria      AT     AUT   
15                                     Azerbaijan      AZ     AZE   
16                                        Bahamas      BS     BHS   
17                                        Bahrain      BH     BHR   
18                                     Bangladesh      BD     BGD   
19                                       Barbados      BB     BRB   
20                                        Belarus      BY     BLR   
21                                        Belgium      BE     BEL   
22                                         Belize      BZ     BLZ   
23                                          Benin      BJ     BEN   
24                                        Bermuda      BM     BMU   
25                                         Bhutan      BT     BTN   
26                                        Bolivia      BO     BOL   
27               Bonaire, Sint Eustatius and Saba      BQ     BES   
28                         Bosnia And Herzegovina      BA     BIH   
29                                       Botswana      BW     BWA   
30                                  Bouvet Island      BV     BVT   
31                                         Brazil      BR     BRA   
32                 British Indian Ocean Territory      IO     IOT   
33                              Brunei Darussalam      BN     BRN   
34                                       Bulgaria      BG     BGR   
35                                   Burkina Faso      BF     BFA   
36                                        Burundi      BI     BDI   
37                                     Cabo Verde      CV     CPV   
38                                       Cambodia      KH     KHM   
39                                       Cameroon      CM     CMR   
40                                         Canada      CA     CAN   
41                                 Cayman Islands      KY     CYM   
42                       Central African Republic      CF     CAF   
43                                           Chad      TD     TCD   
44                                          Chile      CL     CHL   
45                                          China      CN     CHN   
46                               Christmas Island      CX     CXR   
47                        Cocos (Keeling) Islands      CC     CCK   
48                                       Colombia      CO     COL   
49                                        Comoros      KM     COM   
50                                          Congo      CG     COG   
51             Congo (Democratic Republic Of The)      CD     COD   
52                                   Cook Islands      CK     COK   
53                                     Costa Rica      CR     CRI   
54                                  Côte D'Ivoire      CI     CIV   
55                                        Croatia      HR     HRV   
56                                           Cuba      CU     CUB   
57                                        C

In [39]:
# Dropping unnecessary columns
df_clean_countries = df_countries.drop(['alpha-2', 'country-code', 'iso_3166-2', 'sub-region', 'intermediate-region', 'region-code', 'sub-region-code', 'intermediate-region-code'], axis = 1)

In [40]:
df_clean_countries

name alpha-3    region
0                                     Afghanistan     AFG      Asia
1                                   Åland Islands     ALA    Europe
2                                         Albania     ALB    Europe
3                                         Algeria     DZA    Africa
4                                  American Samoa     ASM   Oceania
5                                         Andorra     AND    Europe
6                                          Angola     AGO    Africa
7                                        Anguilla     AIA  Americas
8                                      Antarctica     ATA       NaN
9                             Antigua and Barbuda     ATG  Americas
10                                      Argentina     ARG  Americas
11                                        Armenia     ARM      Asia
12                                          Aruba     ABW  Americas
13                                      Australia     AUS   Oceania
14                                        Austria     AUT    Europe
15                                     Azerbaijan     AZE      Asia
16                                        Bahamas     BHS  Americas
17                                        Bahrain     BHR      Asia
18                                     Bangladesh     BGD      Asia
19                                       Barbados     BRB  Americas
20                                        Belarus     BLR    Europe
21                                        Belgium     BEL    Europe
22                                         Belize     BLZ  Americas
23                                          Benin     BEN    Africa
24                                        Bermuda     BMU  Americas
25                                         Bhutan     BTN      Asia
26                                        Bolivia     BOL  Americas
27               Bonaire, Sint Eustatius and Saba     BES  Americas
28                         Bosnia And Herzegovina     BIH    Europe
29                                       Botswana     BWA    Africa
30                                  Bouvet Island     BVT  Americas
31                                         Brazil     BRA  Americas
32                 British Indian Ocean Territory     IOT    Africa
33                              Brunei Darussalam     BRN      Asia
34                                       Bulgaria     BGR    Europe
35                                   Burkina Faso     BFA    Africa
36                                        Burundi     BDI    Africa
37                                     Cabo Verde     CPV    Africa
38                                       Cambodia     KHM      Asia
39                                       Cameroon     CMR    Africa
40                                         Canada     CAN  Americas
41                                 Cayman Islands     CYM  Americas
42                       Central African Republic     CAF    Africa
43                                           Chad     TCD    Africa
44                                          Chile     CHL  Americas
45                                          China     CHN      Asia
46                               Christmas Island     CXR   Oceania
47                        Cocos (Keeling) Islands     CCK   Oceania
48                                       Colombia     COL  Americas
49                                        Comoros     COM    Africa
50                                          Congo     COG    Africa
51             Congo (Democratic Republic Of The)     COD    Africa
52                                   Cook Islands     COK   Oceania
53                                     Costa Rica     CRI  Americas
54                                  Côte D'Ivoire     CIV    Africa
55                                        Croatia     HRV    Europe
56                                           Cuba     CUB  Americas
57                                        Curaçao     CUW  Americas
58                               

In [41]:
df_bookings.head(10)

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   
5  Resort Hotel            0         14               2015               July   
6  Resort Hotel            0          0               2015               July   
7  Resort Hotel            0          9               2015               July   
8  Resort Hotel            1         85               2015               July   
9  Resort Hotel            1         75               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   
5                        27                          1   
6                        27                          1   
7                        27                          1   
8                        27                          1   
9                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     0       2       0.0       0   
1                        0                     0       2       0.0       0   
2                        0                     1       1       0.0       0   
3                        0                     1       1       0.0       0   
4                        0                     2       2       0.0       0   
5                        0                     2       2       0.0       0   
6                        0                     2       2       0.0       0   
7                        0                     2       2       0.0       0   
8                        0                     3       2       0.0       0   
9                        0                     3       2       0.0       0   

  meal country market_segment distribution_channel  is_repeated_guest  \
0   BB     PRT         Direct               Direct                  0   
1   BB     PRT         Direct               Direct                  0   
2   BB     GBR         Direct               Direct                  0   
3   BB     GBR      Corporate            Corporate                  0   
4   BB     GBR      Online TA                TA/TO                  0   
5   BB     GBR      Online TA                TA/TO                  0   
6   BB     PRT         Direct               Direct                  0   
7   FB     PRT         Direct               Direct                  0   
8   BB     PRT      Online TA                TA/TO                  0   
9   HB     PRT  Offline TA/TO                TA/TO                  0   

   previous_cancellations  previous_bookings_not_canceled reserved_room_type  \
0                       0                               0                  C   
1                       0                               0                  C   
2                       0                               0                  A   
3                       0                               0                  A   
4                       0                               0                  A   
5                       0                               0                  A   
6                       0                               0                  C   
7                       0                               0                  C   
8                       0                               0                  A   
9               

In [42]:
# Dropping Unnecessary columns
df_bookings2 = df_bookings.drop(['reserved_room_type','assigned_room_type','agent','company'], axis = 1)

In [43]:
df_bookings2.head(10)

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   
5  Resort Hotel            0         14               2015               July   
6  Resort Hotel            0          0               2015               July   
7  Resort Hotel            0          9               2015               July   
8  Resort Hotel            1         85               2015               July   
9  Resort Hotel            1         75               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   
5                        27                          1   
6                        27                          1   
7                        27                          1   
8                        27                          1   
9                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     0       2       0.0       0   
1                        0                     0       2       0.0       0   
2                        0                     1       1       0.0       0   
3                        0                     1       1       0.0       0   
4                        0                     2       2       0.0       0   
5                        0                     2       2       0.0       0   
6                        0                     2       2       0.0       0   
7                        0                     2       2       0.0       0   
8                        0                     3       2       0.0       0   
9                        0                     3       2       0.0       0   

  meal country market_segment distribution_channel  is_repeated_guest  \
0   BB     PRT         Direct               Direct                  0   
1   BB     PRT         Direct               Direct                  0   
2   BB     GBR         Direct               Direct                  0   
3   BB     GBR      Corporate            Corporate                  0   
4   BB     GBR      Online TA                TA/TO                  0   
5   BB     GBR      Online TA                TA/TO                  0   
6   BB     PRT         Direct               Direct                  0   
7   FB     PRT         Direct               Direct                  0   
8   BB     PRT      Online TA                TA/TO                  0   
9   HB     PRT  Offline TA/TO                TA/TO                  0   

   previous_cancellations  previous_bookings_not_canceled  booking_changes  \
0                       0                               0                3   
1                       0                               0                4   
2                       0                               0                0   
3                       0                               0                0   
4                       0                               0                0   
5                       0                               0                0   
6                       0                               0                0   
7                       0                               0                0   
8                       0                               0                0   
9                       0           

In [44]:
# Renaming Countries' alpha-3 column to enable merging
df_clean_countries.rename(columns = {'alpha-3' : 'country'}, inplace = True)
df_clean_countries.rename(columns = {'name' : 'country_name'}, inplace = True)

In [45]:
df_clean_countries.head(5)

country_name country   region
0     Afghanistan     AFG     Asia
1   Åland Islands     ALA   Europe
2         Albania     ALB   Europe
3         Algeria     DZA   Africa
4  American Samoa     ASM  Oceania

In [46]:
# Merging clean_countries onto bookings2
df_bookings3 = df_bookings2.merge(df_clean_countries, on = 'country', indicator = True)

In [47]:
df_bookings3.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          0               2015               July   
3  Resort Hotel            0          9               2015               July   
4  Resort Hotel            1         85               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     0       2       0.0       0   
1                        0                     0       2       0.0       0   
2                        0                     2       2       0.0       0   
3                        0                     2       2       0.0       0   
4                        0                     3       2       0.0       0   

  meal country market_segment distribution_channel  is_repeated_guest  \
0   BB     PRT         Direct               Direct                  0   
1   BB     PRT         Direct               Direct                  0   
2   BB     PRT         Direct               Direct                  0   
3   FB     PRT         Direct               Direct                  0   
4   BB     PRT      Online TA                TA/TO                  0   

   previous_cancellations  previous_bookings_not_canceled  booking_changes  \
0                       0                               0                3   
1                       0                               0                4   
2                       0                               0                0   
3                       0                               0                0   
4                       0                               0                0   

  deposit_type  days_in_waiting_list customer_type    adr  \
0   No Deposit                     0     Transient    0.0   
1   No Deposit                     0     Transient    0.0   
2   No Deposit                     0     Transient  107.0   
3   No Deposit                     0     Transient  103.0   
4   No Deposit                     0     Transient   82.0   

   required_car_parking_spaces  total_of_special_requests reservation_status  \
0                            0                          0          Check-Out   
1                            0                          0          Check-Out   
2                            0                          0          Check-Out   
3                            0                          1          Check-Out   
4                            0                          1           Canceled   

  reservation_status_date country_name  region _merge  
0              2015-07-01     Portugal  Europe   both  
1              2015-07-01     Portugal  Europe   both  
2              2015-07-03     Portugal  Europe   both  
3              2015-07-03     Portugal  Europe   both  
4              2015-05-06     Portugal  Europe   both

In [48]:
# Removed _merge column

df_bookings3 = df_bookings3.drop(['_merge'], axis = 1)

In [49]:
df_bookings3.head(5)

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          0               2015               July   
3  Resort Hotel            0          9               2015               July   
4  Resort Hotel            1         85               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     0       2       0.0       0   
1                        0                     0       2       0.0       0   
2                        0                     2       2       0.0       0   
3                        0                     2       2       0.0       0   
4                        0                     3       2       0.0       0   

  meal country market_segment distribution_channel  is_repeated_guest  \
0   BB     PRT         Direct               Direct                  0   
1   BB     PRT         Direct               Direct                  0   
2   BB     PRT         Direct               Direct                  0   
3   FB     PRT         Direct               Direct                  0   
4   BB     PRT      Online TA                TA/TO                  0   

   previous_cancellations  previous_bookings_not_canceled  booking_changes  \
0                       0                               0                3   
1                       0                               0                4   
2                       0                               0                0   
3                       0                               0                0   
4                       0                               0                0   

  deposit_type  days_in_waiting_list customer_type    adr  \
0   No Deposit                     0     Transient    0.0   
1   No Deposit                     0     Transient    0.0   
2   No Deposit                     0     Transient  107.0   
3   No Deposit                     0     Transient  103.0   
4   No Deposit                     0     Transient   82.0   

   required_car_parking_spaces  total_of_special_requests reservation_status  \
0                            0                          0          Check-Out   
1                            0                          0          Check-Out   
2                            0                          0          Check-Out   
3                            0                          1          Check-Out   
4                            0                          1           Canceled   

  reservation_status_date country_name  region  
0              2015-07-01     Portugal  Europe  
1              2015-07-01     Portugal  Europe  
2              2015-07-03     Portugal  Europe  
3              2015-07-03     Portugal  Europe  
4              2015-05-06     Portugal  Europe

In [50]:
df_bookings2.shape

(119390, 28)

In [51]:
df_bookings3.shape

(117620, 30)

In [52]:
# Checking countries of origins
df_bookings2['country'].value_counts(dropna = False).sort_index(ascending=True)

ABW        2
AGO      362
AIA        1
ALB       12
AND        7
ARE       51
ARG      214
ARM        8
ASM        1
ATA        2
ATF        1
AUS      426
AUT     1263
AZE       17
BDI        1
BEL     2342
BEN        3
BFA        1
BGD       12
BGR       75
BHR        5
BHS        1
BIH       13
BLR       26
BOL       10
BRA     2224
BRB        4
BWA        1
CAF        5
CHE     1730
CHL       65
CHN      999
CIV        6
CMR       10
CN      1279
COL       71
COM        2
CPV       24
CRI       19
CUB        8
CYM        1
CYP       51
CZE      171
DEU     7287
DJI        1
DMA        1
DNK      435
DOM       14
DZA      103
ECU       27
EGY       32
ESP     8568
EST       83
ETH        3
FIN      447
FJI        1
FRA    10415
FRO        5
GAB        4
GBR    12129
GEO       22
GGY        3
GHA        4
GIB       18
GLP        2
GNB        9
GRC      128
GTM        4
GUY        1
HKG       29
HND        1
HRV      100
HUN      230
IDN       35
IMN        2
IND      152
IRL     3375

In [53]:
df_bookings3['country'].value_counts(dropna = False).sort_index(ascending=True)

ABW        2
AGO      362
AIA        1
ALB       12
AND        7
ARE       51
ARG      214
ARM        8
ASM        1
ATA        2
ATF        1
AUS      426
AUT     1263
AZE       17
BDI        1
BEL     2342
BEN        3
BFA        1
BGD       12
BGR       75
BHR        5
BHS        1
BIH       13
BLR       26
BOL       10
BRA     2224
BRB        4
BWA        1
CAF        5
CHE     1730
CHL       65
CHN      999
CIV        6
CMR       10
COL       71
COM        2
CPV       24
CRI       19
CUB        8
CYM        1
CYP       51
CZE      171
DEU     7287
DJI        1
DMA        1
DNK      435
DOM       14
DZA      103
ECU       27
EGY       32
ESP     8568
EST       83
ETH        3
FIN      447
FJI        1
FRA    10415
FRO        5
GAB        4
GBR    12129
GEO       22
GGY        3
GHA        4
GIB       18
GLP        2
GNB        9
GRC      128
GTM        4
GUY        1
HKG       29
HND        1
HRV      100
HUN      230
IDN       35
IMN        2
IND      152
IRL     3375
IRN       83

- Bookings 3 Missing CN - China values, TMP - East Timor is now TLS - Timor-Leste, and NaN values
- NaN Values total 488, willing to drop because of no country value
- Change values, update
- Re-add CN and TMP values

In [54]:
df_bookings2['country'] = df_bookings2['country'].replace({'CN': 'CHN'})
df_bookings2['country'] = df_bookings2['country'].replace({'TMP': 'TLS'})

In [55]:
# Re-merging after changes
df_bookings4 = df_bookings2.merge(df_clean_countries, on = 'country', indicator = True)

In [56]:
df_bookings4.shape

(118902, 31)

- 119390 - 488 NaN values = 118902

In [57]:
# Drop _merged column
df_bookings4 = df_bookings4.drop(['_merge'], axis = 1)

In [58]:
# Checking countries of origins
df_bookings4['country'].value_counts(dropna = False).sort_index(ascending=True)

ABW        2
AGO      362
AIA        1
ALB       12
AND        7
ARE       51
ARG      214
ARM        8
ASM        1
ATA        2
ATF        1
AUS      426
AUT     1263
AZE       17
BDI        1
BEL     2342
BEN        3
BFA        1
BGD       12
BGR       75
BHR        5
BHS        1
BIH       13
BLR       26
BOL       10
BRA     2224
BRB        4
BWA        1
CAF        5
CHE     1730
CHL       65
CHN     2278
CIV        6
CMR       10
COL       71
COM        2
CPV       24
CRI       19
CUB        8
CYM        1
CYP       51
CZE      171
DEU     7287
DJI        1
DMA        1
DNK      435
DOM       14
DZA      103
ECU       27
EGY       32
ESP     8568
EST       83
ETH        3
FIN      447
FJI        1
FRA    10415
FRO        5
GAB        4
GBR    12129
GEO       22
GGY        3
GHA        4
GIB       18
GLP        2
GNB        9
GRC      128
GTM        4
GUY        1
HKG       29
HND        1
HRV      100
HUN      230
IDN       35
IMN        2
IND      152
IRL     3375
IRN       83

In [59]:
# Checking for missing Values
df_bookings4.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          4
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
booking_changes                   0
deposit_type                      0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces       0
total_of_special_requests         0
reservation_status                0
reservation_status_date     

In [60]:
df_bookings4[df_bookings4['region'].isnull() == True]

hotel  is_canceled  lead_time  arrival_date_year  \
118887  City Hotel            0         34               2016   
118888  City Hotel            0          7               2017   

       arrival_date_month  arrival_date_week_number  \
118887            October                        42   
118888            January                         3   

        arrival_date_day_of_month  stays_in_weekend_nights  \
118887                         13                        0   
118888                         20                        0   

        stays_in_week_nights  adults  children  babies meal country  \
118887                     2       1       0.0       0   BB     ATA   
118888                     2       1       0.0       0   SC     ATA   

       market_segment distribution_channel  is_repeated_guest  \
118887         Direct               Direct                  0   
118888      Online TA                TA/TO                  0   

        previous_cancellations  previous_bookings_not_canceled  \
118887                       0                               0   
118888                       0                               0   

        booking_changes deposit_type  days_in_waiting_list customer_type  \
118887                0   No Deposit                     0     Transient   
118888                0   No Deposit                     0     Transient   

          adr  required_car_parking_spaces  total_of_special_requests  \
118887  175.5                            0                          0   
118888   93.0                            0                          1   

       reservation_status reservation_status_date country_name region  
118887          Check-Out              2016-10-15   Antarctica    NaN  
118888          Check-Out              2017-01-22   Antarctica    NaN

- Two bookings from Antarctica, which doesn't have a region, is fine.

In [64]:
# Checking for duplicate data
duplicates = df_bookings4[df_bookings4.duplicated()]

In [65]:
duplicates.head(5)

hotel  is_canceled  lead_time  arrival_date_year  \
11   Resort Hotel            0         72               2015   
72   Resort Hotel            1          5               2015   
117  Resort Hotel            1          1               2015   
139  Resort Hotel            0         91               2015   
141  Resort Hotel            0         30               2015   

    arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
11                July                        27                          1   
72                July                        28                          5   
117               July                        28                          8   
139               July                        28                         10   
141               July                        28                         10   

     stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
11                         2                     4       2       0.0       0   
72                         1                     0       2       0.0       0   
117                        0                     1       2       0.0       0   
139                        0                     2       2       0.0       0   
141                        2                     2       2       0.0       0   

    meal country market_segment distribution_channel  is_repeated_guest  \
11    BB     PRT         Direct               Direct                  0   
72    BB     PRT      Online TA                TA/TO                  0   
117   BB     PRT      Corporate            Corporate                  0   
139   BB     PRT      Online TA                TA/TO                  0   
141   FB     PRT         Direct               Direct                  0   

     previous_cancellations  previous_bookings_not_canceled  booking_changes  \
11                        0                               0                1   
72                        0                               0                0   
117                       0                               0                0   
139                       0                               0                0   
141                       0                               0                0   

    deposit_type  days_in_waiting_list    customer_type     adr  \
11    No Deposit                     0        Transient   84.67   
72    No Deposit                     0        Transient   97.00   
117   No Deposit                     0        Transient  104.72   
139   No Deposit                     0        Transient   73.80   
141   No Deposit                     0  Transient-Party  197.00   

     required_car_parking_spaces  total_of_special_requests  \
11                             0                          1   
72                             0                          0   
117                            0                          1   
139                            0                          1   
141                            0                          1   

    reservation_status reservation_status_date country_name  region  
11           Check-Out              2015-07-07     Portugal  Europe  
72            Canceled              2015-07-01     Portugal  Europe  
117           Canceled              2015-07-08     Portugal  Europe  
139          Check-Out              2015-07-12     Portugal  Europe  
141          Check-Out              2015-07-14     Portugal  Europe

In [69]:
df_bookings4.head(15)

hotel  is_canceled  lead_time  arrival_date_year  \
0   Resort Hotel            0        342               2015   
1   Resort Hotel            0        737               2015   
2   Resort Hotel            0          0               2015   
3   Resort Hotel            0          9               2015   
4   Resort Hotel            1         85               2015   
5   Resort Hotel            1         75               2015   
6   Resort Hotel            1         23               2015   
7   Resort Hotel            0         35               2015   
8   Resort Hotel            0         37               2015   
9   Resort Hotel            0         37               2015   
10  Resort Hotel            0         72               2015   
11  Resort Hotel            0         72               2015   
12  Resort Hotel            0         72               2015   
13  Resort Hotel            0         78               2015   
14  Resort Hotel            1         60               2015   

   arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0                July                        27                          1   
1                July                        27                          1   
2                July                        27                          1   
3                July                        27                          1   
4                July                        27                          1   
5                July                        27                          1   
6                July                        27                          1   
7                July                        27                          1   
8                July                        27                          1   
9                July                        27                          1   
10               July                        27                          1   
11               July                        27                          1   
12               July                        27                          1   
13               July                        27                          1   
14               July                        27                          1   

    stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                         0                     0       2       0.0       0   
1                         0                     0       2       0.0       0   
2                         0                     2       2       0.0       0   
3                         0                     2       2       0.0       0   
4                         0                     3       2       0.0       0   
5                         0                     3       2       0.0       0   
6                         0                     4       2       0.0       0   
7                         0                     4       2       0.0       0   
8                         0                     4       2       0.0       0   
9                         0                     4       2       0.0       0   
10                        2                     4       2       0.0       0   
11                        2                     4       2       0.0       0   
12                        2                     4       2       0.0       0   
13                        2                     5       2       0.0       0   
14                        2                     5       2       0.0       0   

   meal country market_segment distribution_channel  is_repeated_guest  \
0    BB     PRT         Direct               Direct                  0   
1    BB     PRT         Direct               Direct                  0   
2    BB     PRT         Direct               Direct                  0   
3    FB     PRT         Direct               Direct                  0   
4    BB     PRT      Online TA                TA/TO                  0   
5    HB     PRT  Offline TA/TO              

In [70]:
df_bookings.head(25)

hotel  is_canceled  lead_time  arrival_date_year  \
0   Resort Hotel            0        342               2015   
1   Resort Hotel            0        737               2015   
2   Resort Hotel            0          7               2015   
3   Resort Hotel            0         13               2015   
4   Resort Hotel            0         14               2015   
5   Resort Hotel            0         14               2015   
6   Resort Hotel            0          0               2015   
7   Resort Hotel            0          9               2015   
8   Resort Hotel            1         85               2015   
9   Resort Hotel            1         75               2015   
10  Resort Hotel            1         23               2015   
11  Resort Hotel            0         35               2015   
12  Resort Hotel            0         68               2015   
13  Resort Hotel            0         18               2015   
14  Resort Hotel            0         37               2015   
15  Resort Hotel            0         68               2015   
16  Resort Hotel            0         37               2015   
17  Resort Hotel            0         12               2015   
18  Resort Hotel            0          0               2015   
19  Resort Hotel            0          7               2015   
20  Resort Hotel            0         37               2015   
21  Resort Hotel            0         72               2015   
22  Resort Hotel            0         72               2015   
23  Resort Hotel            0         72               2015   
24  Resort Hotel            0        127               2015   

   arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0                July                        27                          1   
1                July                        27                          1   
2                July                        27                          1   
3                July                        27                          1   
4                July                        27                          1   
5                July                        27                          1   
6                July                        27                          1   
7                July                        27                          1   
8                July                        27                          1   
9                July                        27                          1   
10               July                        27                          1   
11               July                        27                          1   
12               July                        27                          1   
13               July                        27                          1   
14               July                        27                          1   
15               July                        27                          1   
16               July                        27                          1   
17               July                        27                          1   
18               July                        27                          1   
19               July                        27                          1   
20               July                        27                          1   
21               July                        27                          1   
22               July                        27                          1   
23               July                        27                          1   
24               July                        27                          1   

    stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                         0                     0       2       0.0       0   
1                         0                     0       2       0.0       0   
2                         0                     1       1       0.0       0   
3                         0

In [73]:
df_bookings.shape

(119390, 32)

In [72]:
df_bookings4.shape

(118902, 30)

In [77]:
duplicates.shape

(33006, 30)

In [80]:
# Dropping duplicates, which makes up over 25% of the dataset
df_bookings_nodupes = df_bookings4.drop_duplicates()

In [81]:
df_bookings_nodupes.head(25)

hotel  is_canceled  lead_time  arrival_date_year  \
0   Resort Hotel            0        342               2015   
1   Resort Hotel            0        737               2015   
2   Resort Hotel            0          0               2015   
3   Resort Hotel            0          9               2015   
4   Resort Hotel            1         85               2015   
5   Resort Hotel            1         75               2015   
6   Resort Hotel            1         23               2015   
7   Resort Hotel            0         35               2015   
8   Resort Hotel            0         37               2015   
9   Resort Hotel            0         37               2015   
10  Resort Hotel            0         72               2015   
12  Resort Hotel            0         72               2015   
13  Resort Hotel            0         78               2015   
14  Resort Hotel            1         60               2015   
15  Resort Hotel            0         77               2015   
16  Resort Hotel            0         99               2015   
17  Resort Hotel            1         96               2015   
18  Resort Hotel            1         45               2015   
19  Resort Hotel            1         40               2015   
20  Resort Hotel            0         36               2015   
21  Resort Hotel            1         43               2015   
22  Resort Hotel            1         45               2015   
23  Resort Hotel            0        107               2015   
24  Resort Hotel            1         47               2015   
25  Resort Hotel            0          0               2015   

   arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0                July                        27                          1   
1                July                        27                          1   
2                July                        27                          1   
3                July                        27                          1   
4                July                        27                          1   
5                July                        27                          1   
6                July                        27                          1   
7                July                        27                          1   
8                July                        27                          1   
9                July                        27                          1   
10               July                        27                          1   
12               July                        27                          1   
13               July                        27                          1   
14               July                        27                          1   
15               July                        27                          1   
16               July                        27                          1   
17               July                        27                          1   
18               July                        27                          2   
19               July                        27                          2   
20               July                        27                          2   
21               July                        27                          2   
22               July                        27                          2   
23               July                        27                          2   
24               July                        27                          2   
25               July                        27                          2   

    stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                         0                     0       2       0.0       0   
1                         0                     0       2       0.0       0   
2                         0                     2       2       0.0       0   
3                         0

In [82]:
df_bookings_nodupes.shape

(85896, 30)

In [85]:
# 4 missing values in Children, assumption that no values means no children.
df_bookings_nodupes['children'] = df_bookings_nodupes['children'].fillna(0)

C:\Users\arafo\AppData\Local\Temp\ipykernel_11896\4143271507.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_bookings_nodupes['children'] = df_bookings_nodupes['children'].fillna(0)


In [86]:
df_bookings_nodupes.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
booking_changes                   0
deposit_type                      0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces       0
total_of_special_requests         0
reservation_status                0
reservation_status_date     

In [87]:
df_bookings_nodupes.head(5)

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          0               2015               July   
3  Resort Hotel            0          9               2015               July   
4  Resort Hotel            1         85               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     0       2       0.0       0   
1                        0                     0       2       0.0       0   
2                        0                     2       2       0.0       0   
3                        0                     2       2       0.0       0   
4                        0                     3       2       0.0       0   

  meal country market_segment distribution_channel  is_repeated_guest  \
0   BB     PRT         Direct               Direct                  0   
1   BB     PRT         Direct               Direct                  0   
2   BB     PRT         Direct               Direct                  0   
3   FB     PRT         Direct               Direct                  0   
4   BB     PRT      Online TA                TA/TO                  0   

   previous_cancellations  previous_bookings_not_canceled  booking_changes  \
0                       0                               0                3   
1                       0                               0                4   
2                       0                               0                0   
3                       0                               0                0   
4                       0                               0                0   

  deposit_type  days_in_waiting_list customer_type    adr  \
0   No Deposit                     0     Transient    0.0   
1   No Deposit                     0     Transient    0.0   
2   No Deposit                     0     Transient  107.0   
3   No Deposit                     0     Transient  103.0   
4   No Deposit                     0     Transient   82.0   

   required_car_parking_spaces  total_of_special_requests reservation_status  \
0                            0                          0          Check-Out   
1                            0                          0          Check-Out   
2                            0                          0          Check-Out   
3                            0                          1          Check-Out   
4                            0                          1           Canceled   

  reservation_status_date country_name  region  
0              2015-07-01     Portugal  Europe  
1              2015-07-01     Portugal  Europe  
2              2015-07-03     Portugal  Europe  
3              2015-07-03     Portugal  Europe  
4              2015-05-06     Portugal  Europe

In [88]:
# Renaming adr to average_daily_rate
df_bookings_nodupes.rename(columns = {'adr' : 'average_daily_rate'}, inplace = True)

C:\Users\arafo\AppData\Local\Temp\ipykernel_11896\689024041.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_bookings_nodupes.rename(columns = {'adr' : 'average_daily_rate'}, inplace = True)


In [89]:
df_bookings_nodupes.head(5)

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          0               2015               July   
3  Resort Hotel            0          9               2015               July   
4  Resort Hotel            1         85               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     0       2       0.0       0   
1                        0                     0       2       0.0       0   
2                        0                     2       2       0.0       0   
3                        0                     2       2       0.0       0   
4                        0                     3       2       0.0       0   

  meal country market_segment distribution_channel  is_repeated_guest  \
0   BB     PRT         Direct               Direct                  0   
1   BB     PRT         Direct               Direct                  0   
2   BB     PRT         Direct               Direct                  0   
3   FB     PRT         Direct               Direct                  0   
4   BB     PRT      Online TA                TA/TO                  0   

   previous_cancellations  previous_bookings_not_canceled  booking_changes  \
0                       0                               0                3   
1                       0                               0                4   
2                       0                               0                0   
3                       0                               0                0   
4                       0                               0                0   

  deposit_type  days_in_waiting_list customer_type  average_daily_rate  \
0   No Deposit                     0     Transient                 0.0   
1   No Deposit                     0     Transient                 0.0   
2   No Deposit                     0     Transient               107.0   
3   No Deposit                     0     Transient               103.0   
4   No Deposit                     0     Transient                82.0   

   required_car_parking_spaces  total_of_special_requests reservation_status  \
0                            0                          0          Check-Out   
1                            0                          0          Check-Out   
2                            0                          0          Check-Out   
3                            0                          1          Check-Out   
4                            0                          1           Canceled   

  reservation_status_date country_name  region  
0              2015-07-01     Portugal  Europe  
1              2015-07-01     Portugal  Europe  
2              2015-07-03     Portugal  Europe  
3              2015-07-03     Portugal  Europe  
4              2015-05-06     Portugal  Europe

In [90]:
# Check for mixed types
for col in df_bookings_nodupes.columns.tolist():
  weird = (df_bookings_nodupes[[col]].applymap(type) != df_bookings_nodupes[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_bookings_nodupes[weird]) > 0:
    print (col)

region


- Region has null values because of Antarctica

In [91]:
df_bookings_nodupes['region'] = df_bookings_nodupes['region'].fillna('Antarctica')

C:\Users\arafo\AppData\Local\Temp\ipykernel_11896\1599727519.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_bookings_nodupes['region'] = df_bookings_nodupes['region'].fillna('Antarctica')


In [92]:
# Check for mixed types
for col in df_bookings_nodupes.columns.tolist():
  weird = (df_bookings_nodupes[[col]].applymap(type) != df_bookings_nodupes[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_bookings_nodupes[weird]) > 0:
    print (col)

In [93]:
df_bookings_nodupes.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
booking_changes                   0
deposit_type                      0
days_in_waiting_list              0
customer_type                     0
average_daily_rate                0
required_car_parking_spaces       0
total_of_special_requests         0
reservation_status                0
reservation_status_date     

In [94]:
df_bookings_nodupes.dtypes

hotel                              object
is_canceled                         int64
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                 object
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                               object
country                            object
market_segment                     object
distribution_channel               object
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
booking_changes                     int64
deposit_type                       object
days_in_waiting_list                int64
customer_type                      object
average_daily_rate                

In [97]:
# Change children from float to int
df_bookings_nodupes['children'] = df_bookings_nodupes['children'].astype('int64')

C:\Users\arafo\AppData\Local\Temp\ipykernel_11896\30105098.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_bookings_nodupes['children'] = df_bookings_nodupes['children'].astype('int64')


In [98]:
df_bookings_nodupes['children'].dtypes

dtype('int64')

In [99]:
df_bookings_nodupes.shape

(85896, 30)

In [100]:
# Exporting data to csv
df_bookings_nodupes.to_csv(os.path.join(path, '02 - Data','Prepared Data', 'bookings_nodupes.csv'))

In [101]:
# Exporting as pickle
df_bookings_nodupes.to_pickle(os.path.join(path, '02 - Data','Prepared Data', 'bookings_nodupes.pkl'))